scalableminds/webknossos
# Improve Dataset Settings to not adapt suggestion automatically

This PR changes how wk handles the suggestions of settings in the datasource-properties.json that could be inferred.
In short, it does not automatically apply those suggestions but offers the user to adapt them. It makes it easier for the user if properties in the datasource-properties.json are missing or invalid.

Before you are reading all this. I myself had problems with keeping an overview of all cases and how they should/could be tested and how they should behave. That's why I wrote all this.


There are several cases that might happen. From now on datasource-properties.json will be names datasource.json:
- The datasource.json on the server is correct and the server has no additional things to infer.
   - In that case, nothing needs to be done :).
- The datasource.json on the server is correct and the server has additional things to infer.
   - In that case, the user should be offered to accept these suggestions. Previously, they were adapted directly which might have caused errors that prevented a person from saving other changes done.
- The datasource.json on the server is a correct JSON but has some missing or invalid properties that violate the JSON-schema.
   - In that case, the form will be filled with the invalid JSON and an error should show the user, what in invalid.
   - In case the user does not change the datasource settings, not update will be sent to the backend. This prevents the update request to fail because the datasource settings are still invalid.
- The datasource.json on the server is in an invalid JSON format and therefore cannot be parsed. 
  - If thats the case, then the datasource settings will be filled with the suggested changes by the backend.
  - TODO: Discuss which option to pick:
     - option 1 (Currently the case): Saving the settings is prevented even when the user does not change the datasource settings. Therefore the user is forced to enter valid datasource settings to fix the datasource.json in on the server. 
    - option 2: Saving is not prevented when the datasource settings weren't changed. This makes it possible to e.g. set "Sharing & Permissions"  without having to fix the datasource.
   -> I think we should go for option 1, as ([afaik](https://www.macmillandictionary.com/dictionary/british/afaik)) a dataset with an invalid datasource.json cannot be used anyway.

I hope this covers all cases.



### URL of deployed dev instance (used for testing):
- https://___.webknossos.xyz

### Steps to test:
- Choose one dataset, lets call it A.
- Testing should be done locally as there you can easily change the datasource-properties.json (datasource.json).
- First: Create a backup of the datasource-properties.json of dataset A to recover it after testing. It is easier than using wk's backup file.

#### Testing first case:
> The datasource.json on the server is correct and the server has no additional things to infer

- Open the edit view of dataset A. If it shows a banner that suggestions could be inferred, do the next testing case first.
- If it shows nothing and changing "Sharing & Permissions" works everything is fine :heavy_check_mark: 

#### Testing the second case:
> The datasource.json on the server is correct and the server has additional things to infer.

- To prepare this test case, open the datasource.json of dataset A and remove a resolution from one layer.
- Then go to the dashboard and reload that dataset.
- Then open the edit view. The data tab should now have a blue banner on top that suggests to merge additional findings by the backend into the current settings.
   - First, do not change anything in the datasource settings and change something in one of the other tabs. Saving should work and by reloading/reopening the edit view, the suggested settings should not have been applied and the blue banner should still be there.
  - Next, try to accept the suggested settings and then save the settings. Then reload/reopen the edit view. The blue banner should be gone and the suggested settings should be applied.

#### Testing case three:
> The datasource.json on the server is a correct JSON but has some missing or invalid properties that violate the JSON-schema.

- Open the datasource.json of dataset A, set the largestSegmentId to -1 and remove a resolution again.
- Then reload by using the reload button in the dashboard.
- Next, open the edit view of dataset A. The datasource settings should show the settings correct settings except that the largestSegmentId now is set to -1 and marked as invalid. Additionally, a blue banner should be displayed that shows that the datasource settings have invalid properties. It should have the option of accepting the suggested settings by webknossos. But these suggestions are not helpful in this case, as wk cannot guess the largestSegmentId.
  - First, do the same thing as in the previous test: Try editing something in another tab and the save and do not edit the datasource settings. The saving should work out and be persisted when reloaded.
  - Second, fix the invalid setting by setting the largestSegment to e.g. 100, accept the suggested settings by the backend and then save those settings. This should work fine and the updated settings should be persisted.
  - Next, set the largestSegmentId again to -1 in the datasource.json and reload the settings from the dashboard (Leave the resolutions as they are). 
    The dataset settings should only show a blue banner telling the user to repair the broken settings but the banner does not suggest any improvements from the backend that could be applied.
   Then fix the largestSegmentId again and save the changes. This should work without any error and relaoding the edit page should not show any blue banner anymore.

#### Testing case four:
>The datasource.json on the server is in an invalid JSON format and therefore cannot be parsed. 

- To prepare that case simply remove one comma from the datasource.json of dataset A where a comma is supposed to be. This will break the parsing.
- Reload the dataset from the dashboard to ensure that the datasource-properties.json is up-to-date.
- Then open the edit view. The banner display should be yellow/orange and a warning. The loaded datasource setting should now be the settings that were suggested by the backend. But the backend cannot guess settings like bounding boxes or voxel size. Those settings should be set to "0,0,0,0,0,0" / "0,0,0".
- Try to change something in another tab and then save. Saving should not work and the data tab should open with the incorrect settings.
- Then fix those settings and save them. This should now work and be persisted.


That's it. These were all cases I could think off :)


### Issues:
- fixes #4672

------
